### PR TITLE
Inference: Add the embedding and output layer in the full_iteration_inference cuda graph scope for hybrid models

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Review by kvareddy — APPROVED** [review]

**Comment by tdene** [comment]
Does this work when the CUDA graph scope is something other than `full_iteration_inference`?

If not, could you please leave a comment in the code mentioning this, so that it doesn't trip up future users.

**Comment by tdene** [comment]
This does not capture the `.float()` inside `Float16Module.forward`, nor the `inference_context.last_token_logits` and the `transpose(0, 1).contiguous()` inside the model's `postprocess` calls. I did not check the embedding layer.

Overall, these uncaptured calls add latency and unnecessary memcopies.

References:
 * https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/module.py#L499
 * https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/models/hybrid/hybrid_model.py#L480
 * https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/models/hybrid/hybrid_model.py#L498

EDIT:     
After discussion, the above is wrong. The `hybrid_model.py` calls are captured. What I see in my trace as uncaptured is just the `Float16Module.forward` and a different memcopy (like from `abstract_inference_wrapper`).

Approved; the other sites will be handled by a different PR.

**Review by tdene — APPROVED** [review]

File: megatron/core/models/hybrid/hybrid_model.py [hunk 1]
```diff
@@ -320,6 +322,38 @@ def preprocess_for_fine_grained_offloading(self):
                     off_interface.mark_not_offloadable(param)
             self.disable_param_offloading = False
 
+    def _should_call_local_cudagraph(self, *args, **kwargs):
+        """
+        Check if we should call the local cudagraph path.
+        """
+        if (
+            not self.training
+            and hasattr(self, 'cudagraph_manager')
```
Comment (mathemakitten) [inline megatron/core/models/hybrid/hybrid_model.py:350]: Are we sure that `and kwargs['attention_mask'] is None` is no longer load-bearing? I _think_ it is now superfluous because of the check against inference_context but just double checking.

**Review by mathemakitten — APPROVED** [review]
LGTM. Can you document the speedup?

**Comment by sidsingh-nvidia** [comment]
@mathemakitten we gain around 200-300us with this.

File: megatron/core/models/hybrid/hybrid_model.py (hunk 1 above)
Comment (sidsingh-nvidia) [reply megatron/core/models/hybrid/hybrid_model.py:350]: attention_mask seems like a superfluous check. On this note, we need a standardized way to check if we are in inference_mode. Every module seems to be doing it's own thing atm :). Some check if torch grad is enabled, other's check for the inference context etc.

**Review by shanmugamr1992 — APPROVED** [review]
